File: imagestag/parity/runner.py
"""Python-side parity test runner.

Runs registered parity tests for filters and layer effects,
saves outputs, and compares against JavaScript outputs.

Can also run reference implementations (OpenCV, scikit-image) for
comparison against ImageStag outputs.
"""
from typing import Callable, Any
import numpy as np
from pathlib import Path

from .config import get_test_dir, clear_test_dir
from .comparison import (
    save_test_image,
    compare_outputs,
    save_comparison_image,
    ComparisonResult,
)
from .registry import (
    get_filter_tests,
    get_effect_tests,
    generate_input,
    ParityTestSpec,
    TestCase,
)
from .reference_filters import (
    run_reference_filter,
    save_reference_output,
    list_opencv_filters,
    list_skimage_filters,
)


# Filter function type: (input: ndarray, **params) -> ndarray
FilterFunc = Callable[[np.ndarray], np.ndarray]

# Effect function type: (input: ndarray, **params) -> ndarray
EffectFunc = Callable[[np.ndarray], np.ndarray]

# Registry of filter implementations
_filter_impls: dict[str, FilterFunc] = {}

# Registry of effect implementations
_effect_impls: dict[str, EffectFunc] = {}


def register_filter_impl(name: str, func: FilterFunc) -> None:
    """Register a Python filter implementation for parity testing.

    Args:
        name: Filter name (must match registered parity tests)
        func: Filter function (input: ndarray) -> ndarray
    """
    _filter_impls[name] = func


def register_effect_impl(name: str, func: EffectFunc) -> None:
    """Register a Python effect implementation for parity testing.

    Args:
        name: Effect name (must match registered parity tests)
        func: Effect function (input: ndarray) -> ndarray
    """
    _effect_impls[name] = func


def _strip_f32_suffix(s: str) -> str:
    """Strip _f32 suffix from a string for unified naming."""
    if s.endswith("_f32"):
        return s[:-4]
    return s


def run_filter_test(
    name: str,
    test_case: TestCase,
    save: bool = True,
) -> tuple[np.ndarray, Path | None]:
    """Run a single filter test case.

    Args:
        name: Filter name (may include _f32 suffix)
        test_case: Test case specification
        save: Whether to save the output

    Returns:
        Tuple of (output_image, output_path)
    """
    func = _filter_impls.get(name)
    if func is None:
        raise ValueError(f"No ImageStag implementation registered for filter: {name}")

    # Generate input
    input_img = generate_input(
        test_case.input_generator,
        test_case.width,
        test_case.height,
    )

    # Apply filter (filters should support 1, 3, or 4 channels)
    output = func(input_img)

    # Save if requested
    # Naming: {filter}_{input}_imagestag_{bitdepth}.{format}
    # Strip _f32 suffix from filter name and test_case.id - bit_depth field handles that
    path = None
    if save:
        filter_name = _strip_f32_suffix(name)
        input_name = _strip_f32_suffix(test_case.id)
        path = save_test_image(
            output,
            "filters",
            filter_name,
            input_name,
            "imagestag",
            bit_depth=test_case.bit_depth,
        )

    return output, path


def run_effect_test(
    name: str,
    test_case: TestCase,
    save: bool = True,
) -> tuple[np.ndarray, Path | None]:
    """Run a single effect test case.

    Args:
        name: Effect name (may include _f32 suffix)
        test_case: Test case specification
        save: Whether to save the output

    Returns:
        Tuple of (output_image, output_path)
    """
    func = _effect_impls.get(name)
    if func is None:
        raise ValueError(f"No ImageStag implementation registered for effect: {name}")

    # Generate input
    input_img = generate_input(
        test_case.input_generator,
        test_case.width,
        test_case.height,
    )

    # Apply effect
    output = func(input_img)

    # Save if requested
    # Naming: {effect}_{input}_imagestag_{bitdepth}.{format}
    path = None
    if save:
        effect_name = _strip_f32_suffix(name)
        input_name = _strip_f32_suffix(test_case.id)
        path = save_test_image(
            output,
            "layer_effects",
            effect_name,
            input_name,
            "imagestag",
            bit_depth=test_case.bit_depth,
        )

    return output, path


def run_all_filter_tests(
    name: str | None = None,
    clear: bool = False,
    include_references: bool = True,
) -> dict[str, list[Path]]:
    """Run all registered filter parity tests.

    Also runs OpenCV and scikit-image reference implementations for comparison.

    Args:
        name: If specified, only run tests for this filter
        clear: Whether to clear existing outputs first
        include_references: Whether to also run OpenCV/SKImage references (default: True)

    Returns:
        Dict mapping filter names to list of output paths
    """
    if clear:
        clear_test_dir("filters")

    if name:
        specs = {name: get_filter_tests(name)}
        if specs[name] is None:
            raise ValueError(f"No parity tests registered for filter: {name}")
    else:
        specs = get_filter_tests()

    results: dict[str, list[Path]] = {}

    for filter_name, spec in specs.items():
        if spec is None:
            continue

        results[filter_name] = []
        for test_case in spec.test_cases:
            try:
                _, path = run_filter_test(filter_name, test_case)
                if path:
                    results[filter_name].append(path)

                # Also run reference implementations for u8 tests
                if include_references and test_case.bit_depth == "u8":
                    try:
                        run_reference_tests(filter_name, test_case, save=True)
                    except Exception as ref_e:
                        # Reference tests may fail if library doesn't support the filter
                        pass

            except Exception as e:
                print(f"Error running {filter_name}/{test_case.id}: {e}")

    return results


def run_all_effect_tests(
    name: str | None = None,
    clear: bool = False,
) -> dict[str, list[Path]]:
    """Run all registered effect parity tests.

    Args:
        name: If specified, only run tests for this effect
        clear: Whether to clear existing outputs first

    Returns:
        Dict mapping effect names to list of output paths
    """
    if clear:
        clear_test_dir("layer_effects")

    if name:
        specs = {name: get_effect_tests(name)}
        if specs[name] is None:
            raise ValueError(f"No parity tests registered for effect: {name}")
    else:
        specs = get_effect_tests()

    results: dict[str, list[Path]] = {}

    for effect_name, spec in specs.items():
        if spec is None:
            continue

        results[effect_name] = []
        for test_case in spec.test_cases:
            try:
                _, path = run_effect_test(effect_name, test_case)
                if path:
                    results[effect_name].append(path)
            except Exception as e:
                print(f"Error running {effect_name}/{test_case.id}: {e}")

    return results


def compare_filter_outputs(
    name: str,
    tolerance: float = 0.001,
    save_comparisons: bool = True,
) -> list[ComparisonResult]:
    """Compare Python and JS outputs for a filter.

    Args:
        name: Filter name
        tolerance: Maximum allowed diff ratio
        save_comparisons: Whether to save comparison images for failures

    Returns:
        List of comparison results
    """
    spec = get_filter_tests(name)
    if spec is None:
        raise ValueError(f"No parity tests registered for filter: {name}")

    results = []
    for test_case in spec.test_cases:
        result = compare_outputs("filters", name, test_case.id, tolerance)
        results.append(result)

        if not result.match and save_comparisons:
            save_comparison_image("filters", name, test_case.id)

    return results


# =============================================================================
# Reference Filter Support
# =============================================================================

def run_reference_tests(
    name: str,
    test_case: TestCase,
    save: bool = True,
) -> dict[str, tuple[np.ndarray | None, Path | None]]:
    """Run reference filter implementations for a test case.

    Only runs for u8 test cases (reference libraries don't support f32).

    Args:
        name: Filter name (may include _f32 suffix, will be stripped)
        test_case: Test case specification
        save: Whether to save the outputs

    Returns:
        Dict mapping library names to (output_image, output_path) tuples
    """
    # Skip f32 tests - reference libraries only support 8-bit
    if test_case.bit_depth == "f32":
        return {"opencv": (None, None), "skimage": (None, None)}

    # Strip _f32 suffix for clean naming
    filter_name = _strip_f32_suffix(name)
    input_name = _strip_f32_suffix(test_case.id)

    # Generate input
    input_img = generate_input(
        test_case.input_generator,
        test_case.width,
        test_case.height,
    )

    results = {}

    for library in ["opencv", "skimage"]:
        output = run_reference_filter(filter_name, input_img, test_case.params, library)

        path = None
        if output is not None and save:
            path = save_reference_output(
                output,
                "filters",
                filter_name,
                input_name,
                library,
            )

        results[library] = (output, path)

    return results


def run_all_reference_tests(
    name: str | None = None,
) -> dict[str, dict[str, list[Path]]]:
    """Run all reference implementations for registered filter tests.

    Only runs for u8 (8-bit) test cases.

    Args:
        name: If specified, only run for this filter

    Returns:
        Dict mapping library names to filter name -> paths dict
    """
    if name:
        specs = {name: get_filter_tests(name)}
        if specs[name] is None:
            raise ValueError(f"No parity tests registered for filter: {name}")
    else:
        specs = get_filter_tests()

    results = {"opencv": {}, "skimage": {}}

    for filter_name, spec in specs.items():
        if spec is None:
            continue

        results["opencv"][filter_name] = []
        results["skimage"][filter_name] = []

        for test_case in spec.test_cases:
            # Skip f32 tests
            if test_case.bit_depth == "f32":
                continue

            try:
                ref_outputs = run_reference_tests(filter_name, test_case)

                for library in ["opencv", "skimage"]:
                    _, path = ref_outputs[library]
                    if path:
                        results[library][filter_name].append(path)
            except Exception as e:
                print(f"Error running reference {filter_name}/{test_case.id}: {e}")

    return results


def get_available_references(name: str) -> dict[str, bool]:
    """Check which reference implementations are available for a filter.

    Args:
        name: Filter name

    Returns:
        Dict with "opencv" and "skimage" keys, True if available
    """
    return {
        "opencv": name in list_opencv_filters(),
        "skimage": name in list_skimage_filters(),
    }


def compare_effect_outputs(
    name: str,
    tolerance: float = 0.001,
    save_comparisons: bool = True,
) -> list[ComparisonResult]:
    """Compare Python and JS outputs for an effect.

    Args:
        name: Effect name
        tolerance: Maximum allowed diff ratio
        save_comparisons: Whether to save comparison images for failures

    Returns:
        List of comparison results
    """
    spec = get_effect_tests(name)
    if spec is None:
        raise ValueError(f"No parity tests registered for effect: {name}")

    results = []
    for test_case in spec.test_cases:
        result = compare_outputs("layer_effects", name, test_case.id, tolerance)
        results.append(result)

        if not result.match and save_comparisons:
            save_comparison_image("layer_effects", name, test_case.id)

    return results


class ParityTestRunner:
    """High-level test runner for cross-platform parity tests.

    Example:
        runner = ParityTestRunner()
        runner.register_filter("grayscale", grayscale_filter_func)
        runner.run_python_tests()

        # After JS tests have run...
        report = runner.compare_all()
        print(report)
    """

    def __init__(self, tolerance: float = 0.001):
        """Initialize the test runner.

        Args:
            tolerance: Default tolerance for comparisons
        """
        self.tolerance = tolerance
        self._filter_results: dict[str, list[ComparisonResult]] = {}
        self._effect_results: dict[str, list[ComparisonResult]] = {}

    def register_filter(self, name: str, func: FilterFunc) -> None:
        """Register a filter implementation."""
        register_filter_impl(name, func)

    def register_effect(self, name: str, func: EffectFunc) -> None:
        """Register an effect implementation."""
        register_effect_impl(name, func)

    def run_python_tests(
        self,
        filters: list[str] | None = None,
        effects: list[str] | None = None,
        clear: bool = True,
    ) -> None:
        """Run Python-side tests.

        Args:
            filters: Filter names to test (None = all registered)
            effects: Effect names to test (None = all registered)
            clear: Whether to clear existing outputs first
        """
        if filters is None:
            run_all_filter_tests(clear=clear)
        else:
            for name in filters:
                run_all_filter_tests(name, clear=clear)

        if effects is None:
            run_all_effect_tests(clear=clear)
        else:
            for name in effects:
                run_all_effect_tests(name, clear=clear)

    def compare_all(
        self,
        tolerance: float | None = None,
        save_comparisons: bool = True,
    ) -> dict[str, dict[str, list[ComparisonResult]]]:
        """Compare all Python vs JS outputs.

        Args:
            tolerance: Override default tolerance
            save_comparisons: Save comparison images for failures

        Returns:
            Dict with "filters" and "layer_effects" results
        """
        tol = tolerance or self.tolerance

        filter_tests = get_filter_tests()
        effect_tests = get_effect_tests()

        results = {
            "filters": {},
            "layer_effects": {},
        }

        for name in filter_tests:
            results["filters"][name] = compare_filter_outputs(
                name, tol, save_comparisons
            )

        for name in effect_tests:
            results["layer_effects"][name] = compare_effect_outputs(
                name, tol, save_comparisons
            )

        return results

    def generate_report(
        self,
        results: dict[str, dict[str, list[ComparisonResult]]] | None = None,
    ) -> str:
        """Generate a text report of comparison results.

        Args:
            results: Comparison results (runs compare_all if None)

        Returns:
            Formatted report string
        """
        if results is None:
            results = self.compare_all()

        lines = ["=" * 60, "Cross-Platform Parity Test Report", "=" * 60, ""]

        total_pass = 0
        total_fail = 0

        for category, tests in results.items():
            lines.append(f"\n## {category.upper()}\n")

            for name, comparisons in tests.items():
                lines.append(f"\n### {name}")
                for result in comparisons:
                    status = "✓ PASS" if result.match else "✗ FAIL"
                    lines.append(f"  {status}: {result.message}")
                    if result.match:
                        total_pass += 1
                    else:
                        total_fail += 1

        lines.append("")
        lines.append("=" * 60)
        lines.append(f"Total: {total_pass} passed, {total_fail} failed")
        lines.append("=" * 60)

        return "\n".join(lines)


__all__ = [
    'register_filter_impl',
    'register_effect_impl',
    'run_filter_test',
    'run_effect_test',
    'run_all_filter_tests',
    'run_all_effect_tests',
    'compare_filter_outputs',
    'compare_effect_outputs',
    'run_reference_tests',
    'run_all_reference_tests',
    'get_available_references',
    'ParityTestRunner',
]

File: imagestag/parity/js/runner.js
/**
 * Cross-platform parity test runner for JavaScript.
 *
 * This module runs parity tests for filters and layer effects,
 * saving outputs to the shared temp directory for comparison with Python.
 *
 * Ground truth images are fetched from the API or loaded from pre-generated
 * files to ensure Python and JavaScript use identical inputs.
 *
 * Usage (Node.js):
 *   import { ParityTestRunner } from './runner.js';
 *   const runner = new ParityTestRunner();
 *   runner.registerFilter('grayscale', grayscaleFilter);
 *   await runner.runTests('filters', 'grayscale');
 *
 * Usage (Browser via API):
 *   POST /imgstag/parity/run
 *   { "category": "filters", "name": "grayscale" }
 */

import * as fs from 'fs';
import * as path from 'path';
import { fileURLToPath } from 'url';
import sharp from 'sharp';
import { encode as encodePng } from 'fast-png';

import {
    TEST_WIDTH,
    TEST_HEIGHT,
    TEST_INPUTS,
    DEFAULT_INPUT_NAMES,
} from './constants.js';

// Get project root (this file is at imagestag/parity/js/runner.js)
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const PROJECT_ROOT = path.resolve(__dirname, '..', '..', '..');

// Configuration - uses project-local tmp/ folder
const PARITY_TEST_DIR = path.join(PROJECT_ROOT, 'tmp', 'parity');
const INPUTS_DIR = path.join(PARITY_TEST_DIR, 'inputs');
const OUTPUT_FORMAT = 'png';  // Using PNG for maximum compatibility

// API base URL for fetching ground truth images (when server is running)
const API_BASE_URL = process.env.IMAGESTAG_API_URL || 'http://localhost:8080/imgstag';

/**
 * Load ground truth input from pre-generated .raw file.
 * These files are created by Python's save_ground_truth_inputs().
 *
 * File format: [width: u32] [height: u32] [channels: u32] [data: u8[]]
 *
 * @param {string} inputName - Input name (e.g., 'deer', 'astronaut')
 * @returns {Object|null} - ImageData-like object or null if not found
 */
function loadGroundTruthFromFile(inputName) {
    const inputPath = path.join(INPUTS_DIR, `${inputName}.raw`);

    if (!fs.existsSync(inputPath)) {
        // Try legacy .rgba format
        const legacyPath = path.join(INPUTS_DIR, `${inputName}.rgba`);
        if (fs.existsSync(legacyPath)) {
            const buffer = fs.readFileSync(legacyPath);
            const width = buffer.readUInt32LE(0);
            const height = buffer.readUInt32LE(4);
            const data = new Uint8ClampedArray(buffer.slice(8));
            const channels = data.length / (width * height);
            return { data, width, height, channels };
        }
        return null;
    }

    const buffer = fs.readFileSync(inputPath);
    const width = buffer.readUInt32LE(0);
    const height = buffer.readUInt32LE(4);
    const channels = buffer.readUInt32LE(8);
    const data = new Uint8ClampedArray(buffer.slice(12));

    return { data, width, height, channels };
}

/**
 * Fetch ground truth input from API.
 * Requires the ImageStag server to be running.
 *
 * @param {string} inputName - Input name (e.g., 'deer', 'astronaut')
 * @returns {Promise<Object|null>} - ImageData-like object or null
 */
async function fetchGroundTruthFromAPI(inputName) {
    try {
        const url = `${API_BASE_URL}/parity/inputs/${inputName}.raw`;
        const response = await fetch(url);

        if (!response.ok) {
            console.warn(`Failed to fetch ${inputName} from API: ${response.status}`);
            return null;
        }

        const arrayBuffer = await response.arrayBuffer();
        const buffer = Buffer.from(arrayBuffer);
        const width = buffer.readUInt32LE(0);
        const height = buffer.readUInt32LE(4);
        const channels = buffer.readUInt32LE(8);
        const data = new Uint8ClampedArray(buffer.slice(12));

        return { data, width, height, channels };
    } catch (error) {
        console.warn(`Failed to fetch ${inputName} from API:`, error.message);
        return null;
    }
}

/**
 * Generate test input using a ground truth image.
 *
 * First tries to load from pre-generated file, then falls back to API.
 *
 * @param {string} name - Input name (e.g., 'deer', 'astronaut')
 * @param {number} width - Expected width (for validation)
 * @param {number} height - Expected height (for validation)
 * @returns {Promise<Object>} - ImageData-like object with data, width, height, channels
 */
export async function generateInput(name, width = TEST_WIDTH, height = TEST_HEIGHT) {
    // Get expected channel count from TEST_INPUTS
    const inputConfig = TEST_INPUTS[name];
    const expectedChannels = inputConfig ? inputConfig.channels : 4;

    // First try loading from pre-generated file
    let input = loadGroundTruthFromFile(name);

    // Fall back to API if file not found
    if (!input) {
        input = await fetchGroundTruthFromAPI(name);
    }

    if (!input) {
        throw new Error(
            `Ground truth input '${name}' not found. ` +
            `Run Python tests first to generate inputs, or start the API server.`
        );
    }

    // Validate dimensions
    if (input.width !== width || input.height !== height) {
        console.warn(
            `Input ${name} has unexpected dimensions: ${input.width}x${input.height}, ` +
            `expected ${width}x${height}`
        );
    }

    // Set channels from input or expected
    input.channels = input.channels || expectedChannels;

    return input;
}

/**
 * Strip _f32 suffix from a string for unified naming.
 */
function stripF32Suffix(s) {
    return s.endsWith('_f32') ? s.slice(0, -4) : s;
}

/**
 * Get the output path for a test result.
 *
 * Naming convention: {filter}_{input}_js_{bitdepth}.{format}
 *
 * @param {string} category - "filters" or "layer_effects"
 * @param {string} name - Filter/effect name (may include _f32 suffix)
 * @param {string} testCase - Test case identifier (may include _f32 suffix)
 * @param {string} bitDepth - "u8" or "f32"
 * @param {string} format - Output format
 * @returns {string} - Output file path
 */
export function getOutputPath(category, name, testCase, bitDepth = 'u8', format = null) {
    // Strip _f32 suffixes - bit depth is indicated by the _u8/_f32 suffix instead
    const cleanName = stripF32Suffix(name);
    const cleanTestCase = stripF32Suffix(testCase);

    // f32 always uses PNG for 16-bit precision
    const fmt = format || (bitDepth === 'f32' ? 'png' : OUTPUT_FORMAT);

    const categoryDir = path.join(PARITY_TEST_DIR, category);
    fs.mkdirSync(categoryDir, { recursive: true });
    return path.join(categoryDir, `${cleanName}_${cleanTestCase}_js_${bitDepth}.${fmt}`);
}

/**
 * Create side-by-side comparison image [original | output].
 *
 * @param {Object} inputImage - Original input image
 * @param {Object} outputImage - Processed output image
 * @param {string} bitDepth - "u8" or "f32"
 * @returns {Object} - Combined image data
 */
function createSideBySide(inputImage, outputImage, bitDepth = 'u8') {
    const outWidth = outputImage.width;
    const outHeight = outputImage.height;
    const outChannels = outputImage.channels || 4;

    const inWidth = inputImage.width;
    const inHeight = inputImage.height;
    const inChannels = inputImage.channels || 4;

    // Target channels is max of input/output, at least 3 for visibility
    const targetChannels = Math.max(outChannels, inChannels, 3);

    // Combined width is 2x output width (original is resized/padded to match output)
    const combinedWidth = outWidth * 2;
    const combinedHeight = outHeight;

    // Handle dtype conversion for f32 (12-bit) outputs
    const is12bit = bitDepth === 'f32' && outputImage.data instanceof Uint16Array;
    const maxVal = is12bit ? 4095 : 255;

    // Create combined buffer
    const ArrayType = is12bit ? Uint16Array : Uint8ClampedArray;
    const combined = new ArrayType(combinedWidth * combinedHeight * targetChannels);

    // Helper to ensure image has target channels
    function ensureChannels(img, targetCh, isInput = false) {
        const ch = img.channels || 4;
        const w = img.width;
        const h = img.height;
        const data = img.data;

        // Create result array
        const result = new ArrayType(w * h * targetCh);

        for (let y = 0; y < h; y++) {
            for (let x = 0; x < w; x++) {
                const srcIdx = (y * w + x) * ch;
                const dstIdx = (y * w + x) * targetCh;

                if (ch === 1) {
                    // Grayscale to RGB(A)
                    let val = data[srcIdx];
                    // Convert input u8 to 12-bit if needed
                    if (isInput && is12bit && !(data instanceof Uint16Array)) {
                        val = Math.round(val * 4095 / 255);
                    }
                    result[dstIdx] = val;
                    result[dstIdx + 1] = val;
                    result[dstIdx + 2] = val;
                    if (targetCh === 4) {
                        result[dstIdx + 3] = maxVal;
                    }
                } else if (ch === 3) {
                    // RGB to RGBA
                    for (let c = 0; c < 3; c++) {
                        let val = data[srcIdx + c];
                        if (isInput && is12bit && !(data instanceof Uint16Array)) {
                            val = Math.round(val * 4095 / 255);
                        }
                        result[dstIdx + c] = val;
                    }
                    if (targetCh === 4) {
                        result[dstIdx + 3] = maxVal;
                    }
                } else if (ch === 4) {
                    // RGBA - copy all or truncate to RGB
                    for (let c = 0; c < targetCh; c++) {
                        let val = data[srcIdx + c];
                        if (isInput && is12bit && !(data instanceof Uint16Array)) {
                            val = Math.round(val * 4095 / 255);
                        }
                        result[dstIdx + c] = val;
                    }
                }
            }
        }
        return { data: result, width: w, height: h, channels: targetCh };
    }

    // Prepare input (resize/pad to match output dimensions if needed)
    let preparedInput = ensureChannels(inputImage, targetChannels, true);
    const preparedOutput = ensureChannels(outputImage, targetChannels, false);

    // If input is smaller than output, center it with transparent padding
    if (inWidth < outWidth || inHeight < outHeight) {
        const padded = new ArrayType(outWidth * outHeight * targetChannels);
        const offsetX = Math.floor((outWidth - inWidth) / 2);
        const offsetY = Math.floor((outHeight - inHeight) / 2);

        // Fill with transparent/black background
        if (targetChannels === 4) {
            for (let i = 0; i < padded.length; i += 4) {
                padded[i] = 0;
                padded[i + 1] = 0;
                padded[i + 2] = 0;
                padded[i + 3] = 0; // transparent
            }
        }

        // Copy input to center
        for (let y = 0; y < inHeight; y++) {
            for (let x = 0; x < inWidth; x++) {
                const srcIdx = (y * inWidth + x) * targetChannels;
                const dstIdx = ((y + offsetY) * outWidth + (x + offsetX)) * targetChannels;
                for (let c = 0; c < targetChannels; c++) {
                    padded[dstIdx + c] = preparedInput.data[srcIdx + c];
                }
            }
        }
        preparedInput = { data: padded, width: outWidth, height: outHeight, channels: targetChannels };
    }

    // Combine: [input | output]
    for (let y = 0; y < outHeight; y++) {
        // Left side: input
        for (let x = 0; x < outWidth; x++) {
            const srcIdx = (y * outWidth + x) * targetChannels;
            const dstIdx = (y * combinedWidth + x) * targetChannels;
            for (let c = 0; c < targetChannels; c++) {
                combined[dstIdx + c] = preparedInput.data[srcIdx + c];
            }
        }
        // Right side: output
        for (let x = 0; x < outWidth; x++) {
            const srcIdx = (y * outWidth + x) * targetChannels;
            const dstIdx = (y * combinedWidth + outWidth + x) * targetChannels;
            for (let c = 0; c < targetChannels; c++) {
                combined[dstIdx + c] = preparedOutput.data[srcIdx + c];
            }
        }
    }

    return {
        data: combined,
        width: combinedWidth,
        height: combinedHeight,
        channels: targetChannels
    };
}

/**
 * Save test output image.
 *
 * For 8-bit (u8): saves as lossless AVIF with chromaSubsampling='4:4:4'
 * For 16-bit (f32): saves as 16-bit PNG for cross-platform compatibility
 *
 * Creates side-by-side comparison: [original | output] when inputImage is provided.
 *
 * Output naming: {filter}_{input}_js_{bitdepth}.{format}
 *
 * @param {Object} imageData - Image to save (data, width, height)
 * @param {string} category - Category
 * @param {string} name - Filter/effect name
 * @param {string} testCase - Test case ID
 * @param {string} bitDepth - "u8" for 8-bit or "f32" for 16-bit storage
 * @param {Object|null} inputImage - Original input image for side-by-side comparison
 * @returns {Promise<string>} - Path to saved file
 */
export async function saveTestOutput(imageData, category, name, testCase, bitDepth = 'u8', inputImage = null) {
    // Create side-by-side comparison if input image is provided
    const finalImage = inputImage ? createSideBySide(inputImage, imageData, bitDepth) : imageData;

    // Get channel count from imageData (defaults to 4 if not specified)
    const channels = finalImage.channels || 4;

    if (bitDepth === 'f32' && finalImage.data instanceof Uint16Array) {
        // 16-bit PNG for f32 outputs (cross-platform compatible)
        const outputPath = getOutputPath(category, name, testCase, 'f32');

        // Input data is 12-bit (0-4095), scale to 16-bit range (0-65535)
        const scaled = new Uint16Array(finalImage.data.length);
        for (let i = 0; i < finalImage.data.length; i++) {
            scaled[i] = Math.round(finalImage.data[i] * 65535 / 4095);
        }

        // Use fast-png for proper 16-bit PNG encoding (Sharp doesn't support 16-bit raw input)
        const pngData = encodePng({
            width: finalImage.width,
            height: finalImage.height,
            data: scaled,
            depth: 16,
            channels: channels
        });

        fs.writeFileSync(outputPath, pngData);
        return outputPath;
    } else {
        // 8-bit output as PNG (lossless, cross-platform compatible)
        const outputPath = getOutputPath(category, name, testCase, 'u8', 'png');
        const data = Buffer.from(finalImage.data.buffer);

        await sharp(data, {
            raw: {
                width: finalImage.width,
                height: finalImage.height,
                channels: channels
            }
        }).png().toFile(outputPath);

        return outputPath;
    }
}

/**
 * Load test output from AVIF file.
 *
 * @param {string} category - Category
 * @param {string} name - Filter/effect name
 * @param {string} testCase - Test case ID
 * @returns {Promise<Object|null>} - Loaded image or null
 */
export async function loadTestOutput(category, name, testCase) {
    const outputPath = getOutputPath(category, name, testCase, 'avif');

    if (!fs.existsSync(outputPath)) {
        return null;
    }

    const { data, info } = await sharp(outputPath)
        .ensureAlpha()
        .raw()
        .toBuffer({ resolveWithObject: true });

    return {
        data: new Uint8ClampedArray(data.buffer),
        width: info.width,
        height: info.height
    };
}

/**
 * Parity test runner for JavaScript filters and effects.
 */
export class ParityTestRunner {
    constructor() {
        /** @type {Map<string, Function>} */
        this.filters = new Map();
        /** @type {Map<string, Function>} */
        this.effects = new Map();
        /** @type {Map<string, Object[]>} */
        this.filterTests = new Map();
        /** @type {Map<string, Object[]>} */
        this.effectTests = new Map();
    }

    /**
     * Register a filter implementation.
     *
     * @param {string} name - Filter name
     * @param {Function} func - Filter function (imageData) => imageData
     */
    registerFilter(name, func) {
        this.filters.set(name, func);
    }

    /**
     * Register a layer effect implementation.
     *
     * @param {string} name - Effect name
     * @param {Function} func - Effect function (imageData) => imageData
     */
    registerEffect(name, func) {
        this.effects.set(name, func);
    }

    /**
     * Register test cases for a filter.
     *
     * @param {string} name - Filter name
     * @param {Object[]} testCases - Array of test case objects
     */
    registerFilterTests(name, testCases) {
        this.filterTests.set(name, testCases);
    }

    /**
     * Register test cases for an effect.
     *
     * @param {string} name - Effect name
     * @param {Object[]} testCases - Array of test case objects
     */
    registerEffectTests(name, testCases) {
        this.effectTests.set(name, testCases);
    }

    /**
     * Run tests for a specific filter.
     *
     * @param {string} name - Filter name
     * @returns {Promise<Object[]>} - Results for each test case
     */
    async runFilterTests(name) {
        const func = this.filters.get(name);
        if (!func) {
            throw new Error(`No JS implementation registered for filter: ${name}`);
        }

        const testCases = this.filterTests.get(name) || [];
        const results = [];

        for (const tc of testCases) {
            try {
                const input = await generateInput(tc.inputGenerator, tc.width, tc.height);
                const output = func(input);
                const bitDepth = tc.bitDepth || 'u8';
                // Save pure output for parity testing (no side-by-side)
                const outputPath = await saveTestOutput(output, 'filters', name, tc.id, bitDepth);
                results.push({
                    id: tc.id,
                    success: true,
                    path: outputPath,
                });
            } catch (error) {
                results.push({
                    id: tc.id,
                    success: false,
                    error: error.message,
                });
            }
        }

        return results;
    }

    /**
     * Run tests for a specific effect.
     *
     * @param {string} name - Effect name
     * @returns {Promise<Object[]>} - Results for each test case
     */
    async runEffectTests(name) {
        const func = this.effects.get(name);
        if (!func) {
            throw new Error(`No JS implementation registered for effect: ${name}`);
        }

        const testCases = this.effectTests.get(name) || [];
        const results = [];

        for (const tc of testCases) {
            try {
                const input = await generateInput(tc.inputGenerator, tc.width, tc.height);
                const output = func(input);
                const bitDepth = tc.bitDepth || 'u8';
                // Save pure output for parity testing (no side-by-side)
                const outputPath = await saveTestOutput(output, 'layer_effects', name, tc.id, bitDepth);
                results.push({
                    id: tc.id,
                    success: true,
                    path: outputPath,
                });
            } catch (error) {
                results.push({
                    id: tc.id,
                    success: false,
                    error: error.message,
                });
            }
        }

        return results;
    }

    /**
     * Run all registered tests.
     *
     * @returns {Promise<Object>} - Results organized by category and name
     */
    async runAllTests() {
        const results = {
            filters: {},
            layer_effects: {},
        };

        for (const name of this.filterTests.keys()) {
            if (this.filters.has(name)) {
                results.filters[name] = await this.runFilterTests(name);
            }
        }

        for (const name of this.effectTests.keys()) {
            if (this.effects.has(name)) {
                results.layer_effects[name] = await this.runEffectTests(name);
            }
        }

        return results;
    }
}

// Export configuration for external use
export const config = {
    testDir: PARITY_TEST_DIR,
    inputsDir: INPUTS_DIR,
    outputFormat: OUTPUT_FORMAT,
    apiBaseUrl: API_BASE_URL,
    testInputs: TEST_INPUTS,
};
